lib/macroni/ASTConsumers/Kernel/KernelASTConsumer.cpp

```diff
@@ -95,79 +90,85 @@ void check_releases_function_for_rcu_invocations(vast::hl::FuncOp &func) {
                          "rcu_access_pointer() in critical section\n";
   });
 }
 
-int main(int argc, char **argv) {
-  auto maybe_ast = pasta::parse_ast(argc, argv);
-  if (!maybe_ast.Succeeded()) {
-    std::cerr << maybe_ast.TakeError() << '\n';
-    return EXIT_FAILURE;
-  }
-  auto pasta_ast = maybe_ast.TakeValue();
-
-  // Register the MLIR dialects we will be lowering to
-  mlir::DialectRegistry registry;
-  registry.insert<vast::hl::HighLevelDialect, vast::unsup::UnsupportedDialect,
-                  macroni::macroni::MacroniDialect,
-                  macroni::kernel::KernelDialect>();
-  auto mctx = mlir::MLIRContext(registry);
-
-  // Generate the MLIR
-  auto mod = macroni::generate_macroni_module<KernelCodeGen>(pasta_ast, mctx);
+void KernelASTConsumer::HandleTranslationUnit(clang::ASTContext &Ctx) {
+  // Match safe blocks.
+
+  clang::ast_matchers::MatchFinder finder;
+  rcu_collector matcher;
+  finder.addMatcher(rcu_deference_matcher, &matcher);
+  finder.addMatcher(rcu_deference_bh_matcher, &matcher);
+  finder.addMatcher(rcu_deference_sched_matcher, &matcher);
+  finder.addMatcher(rcu_assign_pointer_matcher, &matcher);
+  finder.addMatcher(rcu_access_pointer_matcher, &matcher);
+  finder.addMatcher(rcu_replace_pointer_matcher, &matcher);
+  finder.matchAST(Ctx);
+
+  // Set up the driver.
+  auto driver =
+      ::macroni::generate_codegen_driver<KernelDialect,
+                                         vast::cg::default_meta_gen>(Ctx);
+
+  driver->push_visitor(std::make_unique<kernel_visitor>(
+      matcher.m_rcu_dereference_to_p, matcher.m_rcu_dereference_bh_to_p,
+      matcher.m_rcu_dereference_sched_to_p,
+      matcher.m_rcu_assign_pointer_to_params, matcher.m_rcu_access_pointer_to_p,
+      matcher.m_rcu_replace_pointer_to_params, Ctx, driver->mcontext(),
+      driver->get_codegen_builder(), driver->get_meta_generator(),
+      driver->get_symbol_generator(),
+      vast::cg::visitor_view(driver->get_visitor_stack())));
+
+  vast::cg::setup_default_visitor_stack(*driver);
+
+  driver->emit(Ctx.getTranslationUnitDecl());
+  driver->finalize();
+  auto mod = driver->freeze();
 
   // Register conversions
-  auto patterns = mlir::RewritePatternSet(&mctx);
-  patterns.add(macroni::kernel::rewrite_get_user)
-      .add(macroni::kernel::rewrite_offsetof)
-      .add(macroni::kernel::rewrite_container_of)
-      .add(macroni::kernel::rewrite_rcu_dereference)
-      .add(macroni::kernel::rewrite_rcu_dereference_check)
-      .add(macroni::kernel::rewrite_rcu_access_pointer)
-      .add(macroni::kernel::rewrite_rcu_assign_pointer)
-      .add(macroni::kernel::rewrite_rcu_replace_pointer)
-      .add(macroni::kernel::rewrite_smp_mb)
-      .add(macroni::kernel::rewrite_list_for_each)
-      .add(macroni::kernel::rewrite_label_stmt)
-      .add(macroni::kernel::rewrite_rcu_read_unlock);
+  auto patterns = mlir::RewritePatternSet(&driver->mcontext());
```

Comment:
Logical split (before you generated, now you are rewriting) -> split into separate components/fns.